src/combinatorics/simple_graph/degree_sum.lean

```diff
@@ -0,0 +1,244 @@
+/-
+Copyright (c) 2020 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kyle Miller.
+-/
+import combinatorics.simple_graph.basic
+import algebra.big_operators.basic
+import data.nat.parity
+import data.zmod.basic
+import tactic.omega
+/-!
+# Degree-sum formula and handshaking lemma
+
+The degree-sum formula is that the sum of the degrees of a finite
+graph is equal to twice the number of edges.  The handshaking lemma is
+a corollary, which is that the number of odd-degree vertices is even.
+
+## Main definitions
+
+- A `dart` is a directed edge, consisting of an ordered pair of adjacent vertices,
+  thought of as being a directed edge.
+- `simple_graph.sum_degrees_eq_twice_card_edges` is the degree-sum formula.
+- `simple_graph.even_card_odd_degree_vertices` is the handshaking lemma.
+- `simple_graph.odd_card_odd_degree_vertices_ne` is that the number of odd-degree
+  vertices different from a given odd-degree vertex is odd.
+- `simple_graph.exists_ne_odd_degree_of_exists_odd_degree` is that the existence of an
+  odd-degree vertex implies the existence of another one.
+
+## Implementation notes
+
+We give a combinatorial proof by using the fact that the map from
+darts to vertices has fibers whose cardinalities are the degrees and
+that the map from darts to edges is 2-to-1.
+
+## Tags
+
+simple graphs, sums, degree-sum formula, handshaking lemma
+-/
+open finset
+
+open_locale big_operators
+
+namespace simple_graph
+universes u
+variables {V : Type u} (G : simple_graph V)
+
+/-- A dart is a directed edge, consisting of an ordered pair of adjacent vertices. -/
+@[ext, derive decidable_eq]
+structure dart :=
+(fst snd : V)
+(is_adj : G.adj fst snd)
+
+/-- There is an equivalence between darts and pairs of a vertex and an incident edge. -/
+@[simps]
+def dart_equiv_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
```

Comment:
Perhaps
```suggestion
def dart.to_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
```

Reply:
I think `equiv` is the right name here, the statement is for an equivalence not a function to sigma.

Reply:
Counter-example: [`linear_map.to_matrix`](https://leanprover-community.github.io/mathlib_docs/linear_algebra/matrix.html#linear_map.to_matrix) is an equiv, and has type `(M₁ →ₗ[R] M₂) ≃ₗ[R] matrix m n R`. #5237 does the same thing.

Reply:
I think those names are exceptions rather than a pattern, and both cases are equivalences of types with extra structure rather than just an equivalence of types.

Reply:
Another example without extra structure is [`multiplicative.to_add`](https://leanprover-community.github.io/mathlib_docs/algebra/group/type_tags.html#multiplicative.to_add)

Reply:
Perhaps a maintainer can weigh in here? @bryangingechen

Reply:
This is a good question. I think recently we have been upgrading lots of functions to `*_equiv`s or `order_iso`s and keeping the original name of the function rather than including `_equiv` or `_order_iso`. 

It's worth getting some more opinions on Zulip (and possibly adding something to our naming conventions).

Reply:
> In most uses, an equivalence is (used as) a function, so it's convenient to name it after what that function does in many cases.

I think this is one of the few cases where it's not meant to be used as a function -- its only purpose is to establish that `G.dart` is a `fintype` (and it would be nicer if `@[derive fintype]` had worked here).  It's incidentally used as a function for `dart.inhabited`, but that instance probably ought to use `dart_of_neighbor_set` anyway.